packages/acceptance-tests/pkg-tests-core/sources/utils/tests.ts

```diff
@@ -166,99 +166,157 @@ exports.startPackageServer = function startPackageServer(): Promise<string> {
+        return;
+      }
 
-    const packStream = fsUtils.packToStream(packageVersionEntry.path, {virtualPath: '/package'});
-    packStream.pipe(res);
+      const packageVersionEntry = packageEntry.get(version);
+      if (!packageVersionEntry) {
+        processError(response, 404, `Package not found: ${name}@${version}`);
+        return;
+      }
 
-    return true;
-  }
+      response.writeHead(200, {
+        ['Content-Type']: 'application/octet-stream',
+        ['Transfer-Encoding']: 'chunked',
+      });
 
-  function processError(res: ServerResponse, statusCode: number, errorMessage: string): boolean {
-    console.error(errorMessage);
+      const packStream = fsUtils.packToStream(packageVersionEntry.path, {virtualPath: '/package'});
+      packStream.pipe(response);
+    },
+  };
 
+  function sendError(res: ServerResponse, statusCode: number, errorMessage: string): void {
     res.writeHead(statusCode);
     res.end(errorMessage);
+  }
 
-    return true;
+  function processError(res: ServerResponse, statusCode: number, errorMessage: string): void {
+    console.error(errorMessage);
+    sendError(res, statusCode, errorMessage);
   }
 
+  function parseRequest(url: string): Request|null {
+    let match: RegExpMatchArray|null;
+
+    url = url.replace(/%2f/g, '/');
+
+    if (match = url.match(/^\/(?:(@[^\/]+)\/)?([^@\/][^\/]*)$/)) {
+      const [_, scope, localName] = match;
+
+      return {
+        type: RequestType.PackageInfo,
+        scope,
+        localName,
+      };
+    } else if (match = url.match(/^\/(?:(@[^\/]+)\/)?([^@\/][^\/]*)\/-\/\2-(.*)\.tgz$/)) {
+      const [_, scope, localName, version] = match;
+
+      return {
+        type: RequestType.PackageTarball,
+        scope,
+        localName,
+        version,
+      };
+    }
+
+    return null;
+  }
+
+  function needsAuth({scope, localName}: Request): boolean {
+    return (scope != null && scope.startsWith('@private')) || localName.startsWith('private');
+  }
+
+  const validAuthorizations = [
+    `Bearer 686159dc-64b3-413e-a244-2de2b8d1c36f`,
+    `Basic dXNlcm5hbWU6YSB2ZXJ5IHNlY3VyZSBwYXNzd29yZA==` // username:a very secure password
+  ];
+
   return new Promise((resolve, reject) => {
     const server = http.createServer(
       (req, res) =>
         void (async () => {
           try {
-            if (await processPackageInfo(req.url.match(/^\/(?:(@[^\/]+)\/)?([^@\/][^\/]*)$/), res))
-              return;
+            const parsedRequest = parseRequest(req.url);
 
-            if (await processPackageTarball(req.url.match(/^\/(?:(@[^\/]+)\/)?([^@\/][^\/]*)\/-\/\2-(.*)\.tgz$/), res))
+            if (parsedRequest == null) {
+              processError(res, 404, `Invalid route: ${req.url}`);
+              return;
+            }
+
+            const {authorization} = req.headers;
+            if (authorization != null) {
+              if (!validAuthorizations.includes(authorization)) {
```

Comment:
I kinda remember the npm registry throwing exceptions when passing authenticated requests to get packages that didn't require authentication - is it still the case? If so, maybe we should add a check against `needsAuth` here (and throw if it returns `false`).

Reply:
It seems to work fine with `--always-auth=true` or when setting `always-auth=true` in my `~/.npmrc`. It also works fine when testin via cURL.

It even works when passing nonsensical tokens via cURL 🤔

```bash
curl -o /dev/null -sSL -D - \
  -H 'authorization: Bearer obviously-not-a-valid-token' \
  https://registry.npmjs.org/typescript/-/typescript-3.4.5.tgz
```

```
HTTP/2 200 
date: Tue, 30 Apr 2019 18:56:13 GMT
content-type: application/octet-stream
content-length: 7876633
set-cookie: __cfduid=dc87b363971ea2df9f6a7947b8622c4241556650573; expires=Wed, 29-Apr-20 18:56:13 GMT; path=/; domain=.registry.npmjs.org; HttpOnly
cf-cache-status: HIT
cache-control: max-age=432000
cf-ray: 4cfbbc06cb91441f-BRU
accept-ranges: bytes
etag: "fc46b3c9fbe7472e55527ec4f16c47a2"
expect-ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
last-modified: Tue, 23 Apr 2019 17:36:18 GMT
vary: Accept-Encoding
server: cloudflare

```

That was the actual command I ran, in other words: I didn't generate a UUID v4 and used that, I literally used `obviously-not-a-valid-token` as token value. The response to

```bash
curl -o /dev/null -sSL -D - \
  -H 'authorization: Basic obviously-not-a-valid-token' \
   https://registry.npmjs.org/typescript/-/typescript-3.4.5.tgz
```

is the same, even though `obviously-not-a-valid-token` isn't a valid base64 string.